tests/Unit/Product/ProductRepositoryTest.php

```diff
@@ -0,0 +1,272 @@
+		$this->assertEquals( [ wc_get_product( $product_1 ) ], $this->product_repository->find( $query_args ) );
+	}
+
+	public function test_find_by_ids() {
+		$product_1 = WC_Helper_Product::create_simple_product();
+		$product_2 = WC_Helper_Product::create_simple_product();
+
+		$ids = [
+			$product_1->get_id(),
+			$product_2->get_id(),
+		];
+
+		$this->assertEqualSets(
+			array_map( 'wc_get_product', $ids ),
+			$this->product_repository->find_by_ids( $ids )
+		);
+	}
+
+	public function test_find_synced_products() {
+		$product_1 = WC_Helper_Product::create_simple_product();
+		$this->product_helper->mark_as_synced( $product_1, $this->generate_google_product_mock() );
+
+		WC_Helper_Product::create_simple_product();
+
+		$this->assertEquals(
+			[ wc_get_product( $product_1->get_id() ) ],
+			$this->product_repository->find_synced_products()
+		);
+
+		$this->assertEquals(
+			[ $product_1->get_id() ],
+			$this->product_repository->find_synced_product_ids()
+		);
+	}
+
+	public function test_find_sync_ready_products() {
+		// create some products that are not sync ready
+
+		// manually set to dont-sync-and-show
+		$product_1 = WC_Helper_Product::create_simple_product();
+		$this->product_meta->update_visibility( $product_1, ChannelVisibility::DONT_SYNC_AND_SHOW );
+
+		// product status is set to private and therefore not supported
+		$product_2 = WC_Helper_Product::create_simple_product();
+		$this->product_meta->update_visibility( $product_2, ChannelVisibility::SYNC_AND_SHOW );
+		$product_2->set_status( 'private' );
+		$product_2->save();
+
+		// product type is not supported
+		$product_3 = WC_Helper_Product::create_external_product();
+		$this->product_meta->update_visibility( $product_3, ChannelVisibility::SYNC_AND_SHOW );
+
+		// syncing this product has failed several times recently
+		$product_4 = WC_Helper_Product::create_simple_product();
+		$this->product_meta->update_visibility( $product_4, ChannelVisibility::SYNC_AND_SHOW );
+		$this->product_meta->update_failed_sync_attempts( $product_4, ProductSyncer::FAILURE_THRESHOLD + 1 );
+		$this->product_meta->update_sync_failed_at( $product_4, strtotime( '+1 year' ) );
+
+		// a simple product that is sync ready
+		$simple_product = WC_Helper_Product::create_simple_product();
+		$this->product_meta->update_visibility( $simple_product, ChannelVisibility::SYNC_AND_SHOW );
+
+		// a variable product that is sync ready along with all its variations
+		$variable_product = WC_Helper_Product::create_variation_product();
+		$this->product_meta->update_visibility( wc_get_product( $variable_product ), ChannelVisibility::SYNC_AND_SHOW );
+		foreach ( $variable_product->get_children() as $variation_id ) {
+			$this->product_meta->update_visibility( wc_get_product( $variation_id ), ChannelVisibility::SYNC_AND_SHOW );
+		}
+
+		$results = $this->product_repository->find_sync_ready_products();
+
+		// compare the IDs because the objects might not be identical
```

Comment:
Why are we able to compare the objects in all the other tests? Is there something specific that's making them different for this test?

Reply:
For some reason, the `internal_meta_keys` property of the `data_store` instance embedded within the `WC_Product` object gets modified here in this particular function. 

It might have something to do with the extra call to `filter_sync_ready_products`, which calls several methods of the `ProductMetaHandler` class, which in turn leads to a call to the WC internal function `maybe_read_meta_data` on the product object. This method is probably adding all those extra data into the product object, making it different than what we expect.

I went with comparing the objects in other tests but generally, I think it's just safer to compare the IDs. Since we don't care about the actual ORM object here.

![image](https://user-images.githubusercontent.com/73110514/124655189-94fab200-deb0-11eb-86b4-1bd8e37b7a13.png)

![image](https://user-images.githubusercontent.com/73110514/124655215-9d52ed00-deb0-11eb-9ab7-60d621ae121b.png)

Reply:
I modified all tests in this test case to only compare product IDs: https://github.com/woocommerce/google-listings-and-ads/pull/855/commits/73a3d5768139921218d4c1b2f964d677887197de

Reply:
Thanks, that's probably more foolproof further down the line anyways.